### PR TITLE
NSFS | NC | Object Tagging

### PR DESCRIPTION
**Pull request by naveenpaul1**
### Explain the changes
1. Tagging should be supported by nsfs
2. Can be stored as simple key-value xattrs with some prefix e.g `user.noobaa.tag.KEY=VALUE`

### Issues: Fixed #xxx / Gap #xxx
1. https://github.com/noobaa/noobaa-core/issues/7454

### Testing Instructions:
1. try to put, get and delete object tag with bellow comands
`aws s3api put-object-tagging --tagging '{"TagSet": [{ "Key": "designation", "Value": "confidential" }, { "Key": "department", "Value": "finance" }, { "Key": "team", "Value": "payroll" } ]}' --bucket {bucke-name} --key {object-key} --endpoint https://localhost:6443 --no-verify-ssl`

`aws s3api get-object-tagging --bucket {bucke-name} --key {object-key}--endpoint https://localhost:6443 --no-verify-ssl`

`aws s3api delete-object-tagging --bucket {bucke-name} --key {object-key} --endpoint https://localhost:6443 --no-verify-ssl`

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/sdk/namespace_fs.js [hunk 1]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
     ////////////////////
 
     async get_object_tagging(params, object_sdk) {
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1782]: The function should also handle versionid, and get the versioned object tagging

File: src/sdk/namespace_fs.js [hunk 2]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
...
+            }
+        } catch (err) {
+            dbg.error(`namespace_fs.get_object_tagging: failed in dir ${file_path} with error: `, err);
+            throw this._translate_object_error_codes(err);
+        }
+        dbg.log0('namespace_fs.get_object_tagging: return tagging ', tag_set, 'file_path :', file_path);
+        return { tagging: tag_set };
     }
+
     async delete_object_tagging(params, object_sdk) {
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1808]: The delete_object_tagging should  handle versionid

File: src/sdk/namespace_fs.js [hunk 3]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
...
+        const file_path = this._get_file_path(params);
+        const fs_context = this.prepare_fs_context(object_sdk);
+        try {
+            await this._clear_user_xattr(fs_context, file_path, XATTR_NOOBAA_CUSTOM_PREFIX);
+        } catch (err) {
+            dbg.error(`namespace_fs.delete_object_tagging: failed in dir ${file_path} with error: `, err);
+            throw this._translate_object_error_codes(err);
+        }
+        //ideal case put should return version id, but nsfs do not support versioning. adding dummy version.
+        return {version_id: 'v1'};
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1819]: I don't think that this is the case, If I recall we do support versioning in NSFS
@romayalon to keep me honest

File: src/sdk/namespace_fs.js [hunk 4]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
...
     async put_object_tagging(params, object_sdk) {
-        // TODO
-        return { tagging: [] };
+        let fs_xattr;
+        const tagging = params.tagging && Object.fromEntries(params.tagging.map(tag => ([tag.key, tag.value])));
+        for (const [xattr_key, xattr_value] of Object.entries(tagging)) {
+            fs_xattr = Object.assign(fs_xattr || {}, {
+                [XATTR_NOOBAA_CUSTOM_PREFIX + xattr_key]: xattr_value,
+            });
+        }
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1829]: I think we can avoid Object.assign by initiating the fs_xattr as an object:
```suggestion
        let fs_xattr = {};
        const tagging = params.tagging && Object.fromEntries(params.tagging.map(tag => ([tag.key, tag.value])));
        for (const [xattr_key, xattr_value] of Object.entries(tagging)) {
              fs_xattr[XATTR_NOOBAA_CUSTOM_PREFIX + xattr_key] = xattr_value;
        }
```
Am I missing something?

File: src/sdk/namespace_fs.js [hunk 5]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
...
+            await this._clear_user_xattr(fs_context, file_path, XATTR_NOOBAA_CUSTOM_PREFIX);
+        } catch (err) {
+            dbg.error(`namespace_fs.delete_object_tagging: failed in dir ${file_path} with error: `, err);
+            throw this._translate_object_error_codes(err);
+        }
+        //ideal case put should return version id, but nsfs do not support versioning. adding dummy version.
+        return {version_id: 'v1'};
     }
+
     async put_object_tagging(params, object_sdk) {
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1822]: Same as in all the ops, I thnk we should handle versions.

File: src/sdk/namespace_fs.js [hunk 6]
```diff
@@ -1777,16 +1780,66 @@ class NamespaceFS {
...
+        try {
+            // remove existng tag before putting new tags
+            await this._clear_user_xattr(fs_context, file_path, XATTR_NOOBAA_CUSTOM_PREFIX);
+            await this.set_fs_xattr_op(fs_context, file_path, fs_xattr, undefined);
+        } catch (err) {
+            dbg.error(`namespace_fs.put_object_tagging: failed in dir ${file_path} with error: `, err);
+            throw this._translate_object_error_codes(err);
+        }
+        //ideal case put should return version id, but nsfs do not support versioning. adding dummy version.
+        return { tagging: [], version_id: 'v1' };
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1842]: Not sure it is the case.

File: src/sdk/namespace_fs.js (hunk 2 above)
Comment (nimrod-becker) [reply src/sdk/namespace_fs.js:1808]: While not part of the next version, it is available upstream and it should be handled in the code, I agree

File: src/sdk/namespace_fs.js (hunk 3 above)
Comment (nimrod-becker) [reply src/sdk/namespace_fs.js:1819]: While not part of the next version, it is available upstream and it should be handled in the code, I agree

File: src/sdk/namespace_fs.js (hunk 5 above)
Comment (nimrod-becker) [reply src/sdk/namespace_fs.js:1822]: While not part of the next version, it is available upstream and it should be handled in the code, I agree

Comment (naveenpaul1) [reply src/sdk/namespace_fs.js:1822]: @liranmauda version changes are handled in latest patch.

File: src/endpoint/s3/ops/s3_put_object_tagging.js [hunk 7]
```diff
@@ -11,7 +11,8 @@ async function put_object_tagging(req, res) {
     const reply = await req.object_sdk.put_object_tagging({
         bucket: req.params.bucket,
         key: req.params.key,
-        tagging: tag_set
+        tagging: tag_set,
+        version_id: req.query.versionId,
```
Comment (liranmauda) [inline src/endpoint/s3/ops/s3_put_object_tagging.js:15]: Actually, I think I missed that one, I think that putting is not supported in was, and we should also not support it.
can you revert the version id handle in put object tagging? 
one thing to notice, is that namespace_s3 does add this into it's params, we should consider removing it from there (as a different task in a different PR).

Comment (naveenpaul1) [reply src/endpoint/s3/ops/s3_put_object_tagging.js]: ok, Will remove version related changes from `put_object_tagging()`

File: src/sdk/namespace_fs.js [hunk 8]
```diff
@@ -1826,16 +1829,73 @@ class NamespaceFS {
...
+            file_path = this._get_file_path(params);
+        }
+        const fs_context = this.prepare_fs_context(object_sdk);
+        try {
+            await this._clear_user_xattr(fs_context, file_path, XATTR_NOOBAA_CUSTOM_PREFIX);
+        } catch (err) {
+            dbg.error(`NamespaceFS.delete_object_tagging: failed in dir ${file_path} with error: `, err);
+            throw this._translate_object_error_codes(err);
+        }
+        //ideal case put should return version id, but nsfs do not support versioning. adding dummy version.
```
Comment (liranmauda) [inline src/sdk/namespace_fs.js:1877]: We can remove this comment

**Review by liranmauda — APPROVED** [review]
LGTM

File: src/sdk/bucketspace_fs.js [hunk 9]
```diff
@@ -247,6 +247,8 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             _.isUndefined(account.nsfs_account_config.gid)) return false;
         try {
             dbg.log0('_has_access_to_nsfs_dir: checking access:', ns.write_resource, account.nsfs_account_config.uid, account.nsfs_account_config.gid);
+            //TODO:  Operation not permitted error on change_user() with non root user, and
```
Comment (romayalon) [inline src/sdk/bucketspace_fs.js:250]: @naveenpaul1 can you explain what is the problem here? this should work with non root user

**Pull request merged**